exceptions thrown during shutdown

# Bug Report

## Describe the Bug

During the shutdown when there's PostgreSQL persistence (e.g. End to end tests) it's very likely to see this stacktrace:
```
[provider-data-plane] SEVERE 2025-04-24T09:23:03.085310191 Runtime exception caught by retry processor: Pool not open
org.eclipse.edc.spi.persistence.EdcPersistenceException: Pool not open
	at org.eclipse.edc.sql.pool.commons.CommonsConnectionPool.getConnection(CommonsConnectionPool.java:71)
	at org.eclipse.edc.sql.datasource.ConnectionPoolDataSource.getConnection(ConnectionPoolDataSource.java:43)
	at org.eclipse.edc.transaction.local.DataSourceResource.getConnection(DataSourceResource.java:89)
	at org.eclipse.edc.sql.store.AbstractSqlStore.getConnection(AbstractSqlStore.java:49)
	at org.eclipse.edc.connector.dataplane.store.sql.SqlDataPlaneStore.lambda$findByIdAndLease$3(SqlDataPlaneStore.java:105)
	at org.eclipse.edc.transaction.local.LocalTransactionContext.execute(LocalTransactionContext.java:74)
	at org.eclipse.edc.connector.dataplane.store.sql.SqlDataPlaneStore.findByIdAndLease(SqlDataPlaneStore.java:104)
	at org.eclipse.edc.statemachine.retry.processor.FutureRetryProcess.lambda$execute$0(FutureRetryProcess.java:53)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalStateException: Pool not open
	at org.apache.commons.pool2.impl.BaseGenericObjectPool.assertOpen(BaseGenericObjectPool.java:453)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:281)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:231)
	at org.eclipse.edc.sql.pool.commons.CommonsConnectionPool.getConnection(CommonsConnectionPool.java:69)
	... 14 more
```

### Expected Behavior

Not to see exceptions

### Observed Behavior

In this specific case it means an ongoing flow cannot be completed correctly during shutdown because the sql pool has been closed beforehand.

## Steps to Reproduce

Run a postgresql backed end to end test, look at the logs

## Possible Implementation

Likely there's the need to reflect about the way the runtime is shut down, there could be multiple solutions here, like:
- adding an additional shutdown phase that is responsible to close "base services" like sql connections or similar
- reintroduce special extension annotations for those who set up base services like the sql pool ones, and they get initialized before/shutdown after the others despite the dependency graph

--- a/core/common/boot/src/main/java/org/eclipse/edc/boot/system/runtime/BaseRuntime.java
+++ b/core/common/boot/src/main/java/org/eclipse/edc/boot/system/runtime/BaseRuntime.java
@@ -39,6 +39,7 @@ import org.jetbrains.annotations.NotNull;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
 
 import static java.lang.Runtime.getRuntime;
 import static java.lang.String.format;
@@ -124,17 +125,8 @@ public class BaseRuntime {
      * shutdown to all the extensions in reverse order to the boot.
      */
     public void shutdown() {
-        var iterator = serviceExtensions.listIterator(serviceExtensions.size());
-        while (iterator.hasPrevious()) {
-            var extension = iterator.previous();
-            try {
-                monitor.debug("Shutdown " + extension.name());
-                extension.shutdown();
-            } catch (Throwable e) {
-                monitor.severe("Error while shutting down extension %s (%s)".formatted(extension.name(), extension.getClass().getSimpleName()), e);
-            }
-        }
-        monitor.info("Shutdown complete");
+        onAllExtensionsInReverseOrder("Shutdown", ServiceExtension::shutdown);
+        onAllExtensionsInReverseOrder("Cleanup", ServiceExtension::cleanup);
     }
 
     protected Monitor getMonitor() {
@@ -199,6 +191,20 @@ public class BaseRuntime {
     @NotNull
     protected Monitor createMonitor() {
         return extensionLoader.loadMonitor(programArgs);
+    }
+
+    private void onAllExtensionsInReverseOrder(String name, Consumer<ServiceExtension> operation) {
+        var iterator = serviceExtensions.listIterator(serviceExtensions.size());
+        while (iterator.hasPrevious()) {
+            var extension = iterator.previous();
+            try {
+                monitor.debug("%s %s".formatted(name, extension.name()));
+                operation.accept(extension);
+            } catch (Throwable e) {
+                monitor.severe("Error during extension %s (%s) %s".formatted(extension.name(), extension.getClass().getSimpleName(), name), e);
+            }
+        }
+        monitor.info("%s complete".formatted(name));
     }
 
 }

--- a/core/common/boot/src/test/java/org/eclipse/edc/boot/system/runtime/BaseRuntimeTest.java
+++ b/core/common/boot/src/test/java/org/eclipse/edc/boot/system/runtime/BaseRuntimeTest.java
@@ -87,10 +87,11 @@ public class BaseRuntimeTest {
     }
 
     @Test
-    void shouldShutdownAllExtensions_whenOneThrowsException() {
+    void shouldShutdownAllExtensionsAndCleanup_whenOneThrowsException() {
         ServiceExtension successful = mock();
         ServiceExtension unsuccessful = mock();
         doThrow(new RuntimeException("shutdown error")).when(unsuccessful).shutdown();
+        doThrow(new RuntimeException("shutdown error")).when(unsuccessful).cleanup();
 
         when(serviceLocator.loadImplementors(eq(ServiceExtension.class), anyBoolean()))
                 .thenReturn(mutableListOf(successful, unsuccessful));
@@ -99,7 +100,9 @@ public class BaseRuntimeTest {
 
         assertThatNoException().isThrownBy(runtime::shutdown);
         verify(successful).shutdown();
+        verify(successful).cleanup();
         verify(unsuccessful).shutdown();
+        verify(unsuccessful).cleanup();
     }
 
     @NotNull

--- a/extensions/common/sql/sql-pool/sql-pool-apache-commons/src/main/java/org/eclipse/edc/sql/pool/commons/CommonsConnectionPoolServiceExtension.java
+++ b/extensions/common/sql/sql-pool/sql-pool-apache-commons/src/main/java/org/eclipse/edc/sql/pool/commons/CommonsConnectionPoolServiceExtension.java
@@ -102,7 +102,7 @@ public class CommonsConnectionPoolServiceExtension implements ServiceExtension {
     }
 
     @Override
-    public void shutdown() {
+    public void cleanup() {
         commonsConnectionPools.forEach(CommonsConnectionPool::close);
     }
 

--- a/spi/common/boot-spi/src/main/java/org/eclipse/edc/spi/system/ServiceExtension.java
+++ b/spi/common/boot-spi/src/main/java/org/eclipse/edc/spi/system/ServiceExtension.java
@@ -27,6 +27,18 @@ public interface ServiceExtension extends SystemExtension {
     }
 
     /**
+     * Hook method to perform some additional preparatory work before the extension is started.
+     * All dependencies are guaranteed to be resolved, and all other extensions are guaranteed to have completed initialization.
+     * <p>
+     * Typical use cases include wanting to wait until all registrations of a {@code *Registry} have completed, perform some additional
+     * checking whether a service exists or not, etc.
+     * <p>
+     * <strong>Do NOT perform any service registration in this method!</strong>
+     */
+    default void prepare() {
+    }
+
+    /**
      * Signals the extension to prepare for the runtime to receive requests.
      */
     default void start() {
@@ -39,14 +51,8 @@ public interface ServiceExtension extends SystemExtension {
     }
 
     /**
-     * Hook method to perform some additional preparatory work before the extension is started.
-     * All dependencies are guaranteed to be resolved, and all other extensions are guaranteed to have completed initialization.
-     * <p>
-     * Typical use cases include wanting to wait until all registrations of a {@code *Registry} have completed, perform some additional
-     * checking whether a service exists or not, etc.
-     * <p>
-     * <strong>Do NOT perform any service registration in this method!</strong>
+     * Do further cleanup of resources that can be still needed during {@link #shutdown()} as database connections, websockets, ...
      */
-    default void prepare() {
+    default void cleanup() {
     }
 }

